### PR TITLE
Add metrics timing message passing from OverseerSubsystemContext to Overseer::route_message

### PR DESCRIPTION
**Pull request by coriolinus**
Closes https://github.com/paritytech/polkadot/issues/2186.

### CODE REVIEW

**Comment by coriolinus** [comment]
Promoting the commit message from 0ab8594:

> This deserves a bit of explanation, as the motivating issue explicitly
> requested randomness. In short, it's hard to get randomness to compile
> for `wasm32-unknown-unknown` because that is explicitly intended to be
> as deterministic as practical. Additionally, even though it would never
> be used for consensus purposes, it still felt offputting to intentionally
> introduce randomness into a node's operations. Except, it wasn't really
> random, either: it was a deterministic PRNG varying only in its state,
> and getting the state to work right for that target would have required
> initializing from a constant.
>
> Given that it was a deterministic sequence anyway, it seemed much simpler
> and more explicit to simply select one of each N messages instead of
> attempting any kind of realistic randomness.

File: node/overseer/src/lib.rs [hunk 1]
```diff
@@ -302,7 +324,83 @@ struct SubsystemInstance<M> {
...
+	tx: mpsc::Sender<MaybeTimed<ToOverseer>>,
+	metrics: Metrics,
+	capture_each: u32,
+	capture_state: u32,
+}
+
+impl<M> OverseerSubsystemContext<M> {
+	/// Create a new `OverseerSubsystemContext`.
+	///
+	/// `capture_each` determines which messages are timed, i.e. a value of 256 means that every 256th
```
Comment (drahnr) [inline node/overseer/src/lib.rs:336]: This is not a good statistical approach. In the good case (which is the case we care about most for timing I assume) a lot of messages have deterministic sequences / follow up messages. 

As such the pick of message should have a true statistical property, such as using a very cheap hash function on `MaybeTimer` creation, i.e. Marsaglia's Xorshift.

At the very least, we should pick a prime for which messages to pick.

File: node/overseer/src/lib.rs [hunk 2]
```diff
@@ -302,7 +324,83 @@ struct SubsystemInstance<M> {
...
+	/// message is timed. A value of `0` has a special meaning: no messages are timed.
```
Comment (drahnr) [inline node/overseer/src/lib.rs:337]: I'd be in favour of either encoding the parameter as en enum with a `From<usize>` implementation for the common case, where the `Disabled` state is explicit. Special meaning numbers tend to become a burden when reading code in the future.

Comment (coriolinus) [reply node/overseer/src/lib.rs]: Resolved in ea59309: went back to the simpler `capture_rate` style, which has no special cases.

File: node/overseer/src/lib.rs (hunk 1 above)
Comment (coriolinus) [reply node/overseer/src/lib.rs]: Agree about the statistics; restored a very simple RNG implementation in ea59309.

**Comment by coriolinus** [comment]
Note to reviewers: the `oorandom::Rand32` implementation added in https://github.com/paritytech/polkadot/pull/2201/commits/ea593098feb8a1de368861be00aab837752551f4 is not suitable for cryptographic purposes. That should be fine because we do not use it for cryptographic purposes.

**Review by drahnr — APPROVED** [review]
LGTM!

**Comment by coriolinus** [comment]
bot merge

**Pull request merged**